apps/gdalalg_vector_geom_simplify_coverage.cpp

```diff
@@ -0,0 +1,220 @@
+        // their geometries to GEOS
+        for (auto &feature : srcLayer)
+        {
+            const OGRGeometry *fgeom = feature->GetGeometryRef();
+
+            // Avoid segfault with non-polygonal inputs on GEOS < 3.12.2
+            // Later versions produce an error instead
+            if (fgeom == nullptr || fgeom->getDimension() != 2)
+            {
+                for (auto &geom : geoms)
+                {
+                    GEOSGeom_destroy_r(m_poGeosContext, geom);
+                }
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "Coverage simplification can only be performed on "
+                         "polygonal geometries.");
+                return false;
+            }
+
+            geoms.push_back(fgeom->exportToGEOS(m_poGeosContext, false));
+
+            feature->SetGeometry(nullptr);  // free some memory
+            feature->SetFDefnUnsafe(dstLayer.GetLayerDefn());
+
+            features.push_back(std::move(feature));
+        }
+
+        // Perform coverage simplifciation
+        GEOSGeometry *coll = GEOSGeom_createCollection_r(
+            m_poGeosContext, GEOS_GEOMETRYCOLLECTION, geoms.data(),
+            static_cast<unsigned int>(geoms.size()));
+
+        if (coll == nullptr)
+        {
+            for (auto &geom : geoms)
+            {
+                GEOSGeom_destroy_r(m_poGeosContext, geom);
+            }
+            return false;
+        }
+
+        GEOSGeometry *geos_result = GEOSCoverageSimplifyVW_r(
+            m_poGeosContext, coll, m_opts.tolerance, m_opts.preserveBoundary);
+        GEOSGeom_destroy_r(m_poGeosContext, coll);
+
+        if (geos_result == nullptr)
+        {
+            return false;
+        }
+
+        m_papoGeosResults = GEOSGeom_releaseCollection_r(
+            m_poGeosContext, geos_result, &m_nGeosResultSize);
+        GEOSGeom_destroy_r(m_poGeosContext, geos_result);
+        CPLAssert(features.size() == m_nGeosResultSize);
+
+        // Create features with the modified geometries
+        for (size_t i = 0; i < features.size(); i++)
+        {
+            GEOSGeometry *dstGeom = m_papoGeosResults[i];
+
+            std::unique_ptr<OGRGeometry> poSimplified(
+                OGRGeometryFactory::createFromGEOS(m_poGeosContext, dstGeom));
+            GEOSGeom_destroy_r(m_poGeosContext, m_papoGeosResults[i]);
+            m_papoGeosResults[i] = nullptr;
+
+            if (poSimplified == nullptr)
+            {
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "Failed to convert result from GEOS");
+                return false;
+            }
+            poSimplified->assignSpatialReference(
+                dstLayer.GetLayerDefn()->GetGeomFieldDefn(0)->GetSpatialRef());
+            features[i]->SetGeometry(std::move(poSimplified));
+
+            if (dstLayer.CreateFeature(features[i].release()) != CE_None)
+            {
+                return false;
+            }
```

Comment:
Using release() will leak as CreateFeature(), perhaps unexpectingly doesn't take ownership of the passed feature.

```suggestion
            if (dstLayer.CreateFeature(features[i].get()) != CE_None)
            {
                return false;
            }
            features[i].reset();
```

Reply:
Fixed, thanks